// bing-wallpaper-api@1.0.34/dist/bing-wallpaper-api.umd.js
(function(global, factory) {
  typeof exports === "object" && typeof module !== "undefined" ? factory(exports, require("dayjs")) : typeof define === "function" && define.amd ? define(["exports", "dayjs"], factory) : (global = typeof globalThis !== "undefined" ? globalThis : global || self, factory(global.BingWallpaperApi = {}, global.dayjs));
})(this, function(exports2, dayjs) {
  "use strict";
  class BingWallpaperError extends Error {
    constructor(message, code) {
      super(message);
      this.code = code;
      this.name = "BingWallpaperError";
    }
  }
  class SimpleCache {
    cache = /* @__PURE__ */ new Map();
    get(key) {
      const item = this.cache.get(key);
      if (!item) {
        return null;
      }
      if (Date.now() - item.timestamp > item.ttl) {
        this.cache.delete(key);
        return null;
      }
      return item.data;
    }
    set(key, data, ttl = 36e5) {
      this.cache.set(key, {
        data,
        timestamp: Date.now(),
        ttl
      });
    }
    clear() {
      this.cache.clear();
    }
  }
  const cache = new SimpleCache();
  function isValidResolution(resolution) {
    const validResolutions = [
      "UHD",
      "1920x1200",
      "1920x1080",
      "1366x768",
      "1280x768",
      "1024x768",
      "800x600",
      "800x480",
      "768x1280",
      "720x1280",
      "640x480",
      "480x800",
      "400x240",
      "320x240",
      "240x320"
    ];
    return validResolutions.includes(resolution);
  }
  function isValidMarket(market) {
    const validMarkets = [
      "zh-CN",
      "en-US",
      "ja-JP",
      "en-AU",
      "en-GB",
      "de-DE",
      "en-NZ",
      "en-CA"
    ];
    return validMarkets.includes(market);
  }
  async function fetchWithTimeout(url, options = {}, timeout = 1e4) {
    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), timeout);
    try {
      const response = await fetch(url, {
        ...options,
        signal: controller.signal
      });
      clearTimeout(id);
      return response;
    } catch (error) {
      clearTimeout(id);
      if (error instanceof Error && error.name === "AbortError") {
        throw new BingWallpaperError("请求超时", "REQUEST_TIMEOUT");
      }
      throw error;
    }
  }
  async function getBingWallpaper(options = {}) {
    const cacheKey = JSON.stringify(options);
    const cachedResult = cache.get(cacheKey);
    if (cachedResult) {
      return cachedResult;
    }
    const {
      date,
      resolution = "1920x1080",
      market = "zh-CN",
      index = 0
    } = options;
    if (index !== void 0 && (index < 0 || index > 7)) {
      throw new BingWallpaperError(`索引必须在 0-7 之间，当前值: ${index}`, "INVALID_INDEX");
    }
    if (resolution && !isValidResolution(resolution)) {
      throw new BingWallpaperError(`不支持的分辨率: ${resolution}`, "INVALID_RESOLUTION");
    }
    if (market && !isValidMarket(market)) {
      throw new BingWallpaperError(`不支持的市场区域: ${market}`, "INVALID_MARKET");
    }
    let targetDate;
    if (date) {
      if (typeof date === "string") {
        targetDate = dayjs(date);
        if (!targetDate.isValid()) {
          throw new BingWallpaperError("无效的日期格式", "INVALID_DATE");
        }
      } else if (date instanceof Date) {
        targetDate = dayjs(date);
        if (!targetDate.isValid()) {
          throw new BingWallpaperError("无效的日期对象", "INVALID_DATE");
        }
      } else {
        targetDate = date;
        if (!targetDate.isValid()) {
          throw new BingWallpaperError("无效的日期对象", "INVALID_DATE");
        }
      }
    } else {
      targetDate = dayjs();
    }
    let calculatedIndex = index;
    if (date) {
      const today = dayjs();
      calculatedIndex = today.diff(targetDate, "day");
      if (calculatedIndex < 0 || calculatedIndex > 7) {
        throw new BingWallpaperError(
          `索引超出范围，仅支持0-7之间的值，当前值: ${calculatedIndex}`,
          "INDEX_OUT_OF_RANGE"
        );
      }
    }
    try {
      const apiUrl = `https://www.bing.com/HPImageArchive.aspx?format=js&idx=${calculatedIndex}&n=1&mkt=${market}`;
      const response = await fetchWithTimeout(apiUrl);
      if (!response.ok) {
        throw new BingWallpaperError(
          `Bing API 请求失败，状态码: ${response.status} ${response.statusText}`,
          "API_ERROR"
        );
      }
      const data = await response.json();
      if (!data.images || data.images.length === 0) {
        throw new BingWallpaperError("没有找到壁纸数据", "NO_DATA_FOUND");
      }
      const image = data.images[0];
      let imageUrl = image.url;
      if (!imageUrl.startsWith("http")) {
        imageUrl = `https://www.bing.com${imageUrl}`;
      }
      if (resolution !== "1920x1080") {
        imageUrl = imageUrl.replace(/1920x1080/, resolution);
      }
      const result = {
        url: imageUrl,
        title: image.title || "必应每日壁纸",
        copyright: image.copyright || "",
        copyrightlink: image.copyrightlink ? image.copyrightlink.startsWith("http") ? image.copyrightlink : `https://www.bing.com${image.copyrightlink}` : void 0,
        startdate: image.startdate || targetDate.format("YYYYMMDD"),
        urlbase: image.urlbase || ""
      };
      cache.set(cacheKey, result);
      return result;
    } catch (error) {
      try {
        const backupUrl = "https://bingw.jasonzeng.dev/?format=json";
        const response = await fetchWithTimeout(backupUrl);
        if (!response.ok) {
          throw new BingWallpaperError(
            `备用 API 请求失败，状态码: ${response.status} ${response.statusText}`,
            "BACKUP_API_ERROR"
          );
        }
        const data = await response.json();
        const result = {
          url: data.url || "",
          title: data.title || "必应每日壁纸",
          copyright: data.copyright || "",
          copyrightlink: void 0,
          startdate: data.startdate || dayjs().format("YYYYMMDD"),
          urlbase: data.urlbase || ""
        };
        cache.set(cacheKey, result);
        return result;
      } catch (backupError) {
        if (error instanceof BingWallpaperError) {
          throw error;
        }
        throw new BingWallpaperError(
          `获取必应壁纸失败: ${error instanceof Error ? error.message : "未知错误"}`,
          "FETCH_FAILED"
        );
      }
    }
  }
  exports2.BingWallpaperError = BingWallpaperError;
  exports2.default = getBingWallpaper;
  exports2.getBingWallpaper = getBingWallpaper;
  Object.defineProperties(exports2, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
});
